Read/write separation for procedures.

jakewins opened the pull request:
This utilizes the KernelTX `AccessMode` to set the current transaction in read-only mode when invoking read-only procedures. It also introduces a separate mode to invoke writing procedures, sitting on `DataWriteOperations`. 

`ProcedureSignature` now includes a mode that encodes the execution mode of a procedure, and the cypher runtime uses this to choose which of the two kernel primitives to use to invoke the procedure.

Mats-SX commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/procs/ProcedureExecutionResult.scala:
```diff
@@ -48,7 +48,8 @@ case class ProcedureExecutionResult[E <: Exception](taskCloser: TaskCloser,
   override def accept[EX <: Exception](visitor: InternalResultVisitor[EX]) = {
-    context.callReadOnlyProcedure(signature, args.map(asJavaCompatible)).foreach { res =>
+    val procedure: Iterator[Array[AnyRef]] = signature.mode.call(context, signature, args.map(asJavaCompatible))
+    procedure.foreach { res =>
```
Could be inlined.

Mats-SX commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/procs/ProcedureExecutionResult.scala:
```diff
@@ -60,10 +61,13 @@ case class ProcedureExecutionResult[E <: Exception](taskCloser: TaskCloser,
+
+  override def close() = {
+    taskCloser.close(success = true)
+  }
```
Could remove braces.

Mats-SX commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/spi/ProcedureSignature.scala:
```diff
@@ -19,11 +19,28 @@
+  override val queryType: InternalQueryType = READ_WRITE
+  override def call(ctx:QueryContext, signature: ProcedureSignature, args: Seq[Any]): Iterator[Array[AnyRef]] =
+    ctx.callReadWriteProcedure(signature, args)
+}
```
Please style-format these three classes.

Mats-SX commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/spi/ProcedureSignature.scala:
```diff
@@ -19,11 +19,28 @@
+  val queryType : InternalQueryType
+  def call(ctx:QueryContext, signature:ProcedureSignature, args:Seq[Any]) : Iterator[Array[AnyRef]]
+}
+case object PROC_READ_ONLY extends ProcedureMode {
```
Scala conventions for `object`s is camelCase; `procReadOnly`, but this could be an exception if we want to design procedures as a visibly different entity.

Mats-SX commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/ExceptionTranslatingQueryContextFor3_0.scala:
```diff
@@ -125,6 +127,10 @@ class ExceptionTranslatingQueryContextFor3_0(inner: QueryContext) extends Delega
 
+  override def callReadWriteProcedure(signature: ProcedureSignature, args: Seq[Any]): Iterator[Array[AnyRef]] = {
+    translateIterator(super.callReadWriteProcedure(signature, args))
+  }
```
Could remove braces.

Mats-SX commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_0/TransactionBoundPlanContext.scala:
```diff
@@ -127,8 +127,14 @@ class TransactionBoundPlanContext(initialStatement: Statement, val gdb: GraphDat
+    ProcedureSignature(name, input, output, mode)
+  }
+
+  private def asCypherProcMode( mode:KernelProcedureSignature.Mode ): ProcedureMode = mode match {
```
Wrong java-style spacing of arguments.

Mats-SX commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_0/TransactionBoundPlanContext.scala:
```diff
@@ -127,8 +127,14 @@ class TransactionBoundPlanContext(initialStatement: Statement, val gdb: GraphDat
+
+  private def asCypherProcMode( mode:KernelProcedureSignature.Mode ): ProcedureMode = mode match {
+    case KernelProcedureSignature.Mode.READ_ONLY => PROC_READ_ONLY
+    case KernelProcedureSignature.Mode.READ_WRITE => PROC_READ_WRITE
```
This is an exhaustive match currently, but perhaps it's a good idea to add a default case anyway, for a future implementor of another mode? Ie, `case _ => throw new IllegalArgumentException("unknown mode")`.

jakewins committed:
Read/Write separation of procedures.

This utilizes the KernelTX `AccessMode` to set the current transaction in
read-only mode when invoking read-only procedures. It also introduces a
separate mode to invoke writing procedures, sitting on
`DataWriteOperations`.

`ProcedureSignature` now includes a mode that encodes the execution mode of
a procedure, and the cypher runtime uses this to choose which of the two
kernel primitives to use to invoke the procedure.

In order to support side-effect-only procedures in a helpful way,
this commit also allows procedures to return Void.

jakewins committed:
Add `CALL` to cypher documentation

jakewins committed:
Handle procedures in root package, and improve error message when using stdlib types as records.

jakewins commented:
@Mats-SX @boggle think I've addressed all your comments now! Let me know what you think, if we can get this merged that'd be brilliant because it'd unblock some work @jjaderberg and I want to do in docs

Mats-SX commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_0/TransactionBoundPlanContext.scala:
```diff
@@ -127,8 +127,16 @@ class TransactionBoundPlanContext(initialStatement: Statement, val gdb: GraphDat
+    case KernelProcedureSignature.Mode.READ_ONLY => ProcReadOnly
+    case KernelProcedureSignature.Mode.READ_WRITE => ProcReadWrite
+    case _ => throw new CypherExecutionException(
+      "Unable to execute procedure, because it requires an unrecognized excution mode: " + mode.name(), null )
```
Typo: `excution`

Mats-SX commented:
@jakewins Looks good, just one small textual issue that I found.

jakewins commented:
Merging this since both Mats and Stefan have signed off on it.
